coordinator/tributary/src/mempool.rs

```diff
@@ -4,108 +4,165 @@ use ciphersuite::{Ciphersuite, Ristretto};
 }
 
-impl<D: Db, T: Transaction> Mempool<D, T> {
+impl<D: Db, T: TransactionTrait> Mempool<D, T> {
   fn transaction_key(&self, hash: &[u8]) -> Vec<u8> {
     D::key(b"tributary_mempool", b"transaction", [self.genesis.as_ref(), hash].concat())
   }
   fn current_mempool_key(&self) -> Vec<u8> {
     D::key(b"tributary_mempool", b"current", self.genesis)
   }
 
+  /// save given tx to the mempool db
+  fn save_tx(&mut self, tx: Transaction<T>) {
+    let tx_hash = tx.hash();
+    let transaction_key = self.transaction_key(&tx_hash);
+    let current_mempool_key = self.current_mempool_key();
+    let mut current_mempool = self.db.get(&current_mempool_key).unwrap_or(vec![]);
+
+    let mut txn = self.db.txn();
+    txn.put(transaction_key, tx.serialize());
+    current_mempool.extend(tx_hash);
+    txn.put(current_mempool_key, current_mempool);
+    txn.commit();
+
+    self.txs.insert(tx_hash, tx);
+  }
+
+  fn unsigned_already_exist(&self, unsigned_included: &[[u8; 32]], hash: [u8; 32]) -> bool {
+    // check we have the tx in the pool/chain
+    let res = unsigned_included.iter().find(|&&tx_hash| {
+      tx_hash == hash
+    });
+    if res.is_some() || self.txs.contains_key(&hash){
+      return true;
+    }
+
+    false
+  }
+
   pub(crate) fn new(db: D, genesis: [u8; 32]) -> Self {
-    let mut res = Mempool { db, genesis, txs: HashMap::new(), next_nonces: HashMap::new() };
+    let mut res = Mempool { db, genesis, txs: HashMap::<[u8; 32], Transaction<T>>::new(), next_nonces: HashMap::new() };
 
     let current_mempool = res.db.get(res.current_mempool_key()).unwrap_or(vec![]);
     let mut hash = [0; 32];
     let mut i = 0;
     while i < current_mempool.len() {
       hash.copy_from_slice(&current_mempool[i .. (i + 32)]);
-      let tx =
-        T::read::<&[u8]>(&mut res.db.get(res.transaction_key(&hash)).unwrap().as_ref()).unwrap();
-
-      match tx.kind() {
-        TransactionKind::Signed(Signed { signer, nonce, .. }) => {
-          if let Some(prev) = res.next_nonces.insert(*signer, nonce + 1) {
-            // These mempool additions should've been ordered
-            assert!(prev < *nonce);
+      let tx: Transaction<T> =
+        Transaction::read::<&[u8]>(&mut res.db.get(res.transaction_key(&hash)).unwrap().as_ref()).unwrap();
+
+      debug_assert_eq!(tx.hash(), hash);
+      match tx {
+        Transaction::Tendermint(tx) => {
+          res.txs.insert(hash, Transaction::Tendermint(tx));
+        },
+        Transaction::Application(tx) => {
+          match tx.kind() {
+            TransactionKind::Signed(Signed { signer, nonce, .. }) => {
+              if let Some(prev) = res.next_nonces.insert(*signer, nonce + 1) {
+                // These mempool additions should've been ordered
+                assert!(prev < *nonce);
+              }
+              res.txs.insert(hash, Transaction::Application(tx));
+            }
+            _ => panic!("mempool database had a non-signed application transaction"),
```

Comment:
Unsigned transactions are now mempool routed, so we need to handle Transaction::Unsigned and update this error to "non-provided".

Reply:
@akildemir Application transactions can be unsigned and are allowed in the mempool so long as they're valid.